### PR TITLE
Add multiple quantity consumables checkout

### LINKED ISSUE #1820: Feature Request: Allow Consumables Checkout Quantity to be Specified
FEATURE REQUEST
## Not all consumables (like Ethernet cables) are used one at a time.  A higher quantity of cables may be used or checked out.  If that number is significant, using the "Checkout" feature is a daunting task, requiring the user to either checkout many times or just manually edit the on-hand count.

There is no quantity option.  Therefore; a user must click "Checkout" as many times as they need to reflect the quantity of consumable they are taking.

---
#### Please provide answers to these questions before posting your bug report:
- Version of Snipe-IT you're running 
  v2.1.0-41
- What OS and web server you're running Snipe-IT on 
  CentOS 6.6 / Apache
- What method you used to install Snipe-IT (install.sh, manual installation, docker, etc) 
  install.sh
- If you're getting an error in your browser, include that error 
  N/A - Feature Request
- What specific Snipe-IT page you're on, and what specific element you're interacting with to trigger the error
  /admin/consumables
- If a stacktrace is provided in the error, include that too.
  N/A - Feature Request
- Any errors that appear in your browser's error console.
  N/A - Feature Request
- Confirm whether the error is [reproduceable on the demo](https://snipeitapp.com/demo).
  Yes
- Include any additional information you can find in `app/storage/logs` and your webserver's logs.
  N/A - Feature Request
- Include what you've done so far in the installation, and if you got any error messages along the way.
  N/A - Feature Request
- Indicate whether or not you've manually edited any data directly in the database
  No

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/31590702-feature-request-allow-consumables-checkout-quantity-to-be-specified?utm_campaign=plugin&utm_content=tracker%2F505106&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F505106&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by veenone**
# Description

This branch extends checkout functionality for consumable items to support multiple number of item per check out.
Each checkout will have its own description for historical purpose.
Remaining item calculation is now based on total item checked out

Fixes #1820 #5176

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Test A : validating all functions working as intended and mitigate implementation side effects (other view using the updated data. i.e : consumables view in user)
- [x] Test B : validating on error handler during quantity comparison between total stock, remaining and checked out stock

**Test Configuration**:
* PHP version: 7.4.27
* MySQL version : 10.4.22
* Webserver version: Apache 2.4.52
* OS version: Windows 10

# Checklist:

- [x] I have read the Contributing documentation available here: https://snipe-it.readme.io/docs/contributing-overview
- [x] I have formatted this PR according to the project guidelines: https://snipe-it.readme.io/docs/contributing-overview#pull-request-guidelines
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

# screens

## consumable checkout
checkout-consumables

## consumable details
consumable-details

## consumable stock validation
consumable-stock checking

## user's consumable view
user-consumable

# Remarks
I'm sorry I don't have a knowledge on web programming let alone PHP & laravel mvc implementation.
I'm open for any suggestion & guidance 🙇‍♂️

### CODE REVIEW

File: app/Http/Controllers/Consumables/ConsumableCheckoutController.php [hunk 1]
```diff
@@ -48,6 +49,14 @@ public function store(Request $request, $consumableId)
             return redirect()->route('consumables.index')->with('error', trans('admin/consumables/message.not_found'));
         }
 
+        $request->validate([
+            "totalnum" => "required|regex:/^[0-9]*$/|max:{{$consumable->numRemaining()}}"
```
Comment (snipe) [inline app/Http/Controllers/Consumables/ConsumableCheckoutController.php]: Blade syntax is only for blades - I don’t think this will work here

Comment (snipe) [reply app/Http/Controllers/Consumables/ConsumableCheckoutController.php]: (Even if it did, we’d want a custom validator here, so we could localize the strings)

File: database/migrations/2013_11_15_190357_create_licenses_table.php [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+use Illuminate\Database\Migrations\Migration;
+
+class CreateLicensesTable extends Migration
+{
+    /**
+     * Run the migrations.
+     *
+     * @return void
+     */
+    public function up()
```
Comment (snipe) [inline database/migrations/2013_11_15_190357_create_licenses_table.php]: Why is this here? This is a migration from 2013

Comment (veenone) [reply database/migrations/2013_11_15_190357_create_licenses_table.php]: this one is a mistake from me. I mistakenly added the file because I didn't fast forward the repo after initial forking of the repo.

To be deleted

File: app/Http/Controllers/Consumables/ConsumableCheckoutController.php (hunk 1 above)
Comment (veenone) [reply app/Http/Controllers/Consumables/ConsumableCheckoutController.php]: any advice you might have for the proper validation syntax location?

Comment (veenone) [reply app/Http/Controllers/Consumables/ConsumableCheckoutController.php]: string will be updated to support localization

Comment (veenone) [reply app/Http/Controllers/Consumables/ConsumableCheckoutController.php]: for now,  strings on validator has been moved to admin/consumables/message.
so far checking for these 3 validations work. but whenever I switch my locale to different language,  it doesn't switch. strange..

File: app/Events/CheckoutableCheckedOut.php [hunk 3]
```diff
@@ -14,17 +14,24 @@ class CheckoutableCheckedOut
     public $checkedOutTo;
     public $checkedOutBy;
     public $note;
+    public $total;
```
Comment (uberbrady) [inline app/Events/CheckoutableCheckedOut.php:18]: The attribute seems to be called `$totalnum` in some places, and `$total` here. Also I think we would have to add the `checkoutnote` field as well?

File: app/Models/Consumable.php [hunk 4]
```diff
@@ -285,13 +286,87 @@ public function getEula()
      */
     public function numRemaining()
     {
+        $checkedouttotal = null;
         $checkedout = $this->users->count();
+        foreach($this->users as $data){
+                $totalnum = $data['pivot']['totalnum'];
+                Log:info($totalnum);
```
Comment (uberbrady) [inline app/Models/Consumable.php]: This should probably be `Log::debug()` (two colons, and debug level).

File: app/Models/Consumable.php [hunk 5]
```diff
@@ -285,13 +286,87 @@ public function getEula()
      */
     public function numRemaining()
     {
+        $checkedouttotal = null;
         $checkedout = $this->users->count();
+        foreach($this->users as $data){
+                $totalnum = $data['pivot']['totalnum'];
```
Comment (uberbrady) [inline app/Models/Consumable.php]: This may work as-is, but we prefer to refer to this as: `$data->pivot->totalnum`

File: app/Models/Consumable.php [hunk 6]
```diff
@@ -285,13 +286,87 @@ public function getEula()
      */
     public function numRemaining()
     {
+        $checkedouttotal = null;
         $checkedout = $this->users->count();
+        foreach($this->users as $data){
```
Comment (uberbrady) [inline app/Models/Consumable.php]: Usually variable names like `$data` are a bad idea - since this is a list of users, how about just `$user`?

File: app/Models/Consumable.php [hunk 7]
```diff
@@ -285,13 +286,87 @@ public function getEula()
...
 
+    
+    /**
+     * Gets the number of consumed consumables
+     *
+     * @author [A. Rahardianto] [<veenone@gmail.com>]
+     * @since [v.6]
+     * @return int
+     */
+    public function numConsumed()
```
Comment (uberbrady) [inline app/Models/Consumable.php:319]: This looks a bit like copy-paste code from `numRemaining`, above. Can we do something either there or here that does `$total - $this->numConsumed()` so we don't duplicate the logic?

File: app/Models/Consumable.php [hunk 8]
```diff
@@ -285,13 +286,87 @@ public function getEula()
...
+     *
+     * @author [A. Rahardianto] [<veenone@gmail.com>]
+     * @since [v.6]
+     * @return int
+     */
+    public function lastConsumed()
+    {
+        $checkedouttotal = null;      
+        $checkoutnote = $this->users()->orderBy('created_at','desc')->get();
+        $last = $checkoutnote->last()['pivot']['totalnum'];
```
Comment (uberbrady) [inline app/Models/Consumable.php]: Same note about the `pivot` here - also can we change the name on this function? Something a little more clear like `lastConsumedQty()` or something?

File: app/Models/Consumable.php [hunk 9]
```diff
@@ -285,13 +286,87 @@ public function getEula()
...
+    /**
+     * Gets the number of checked out consumables
+     *
+     * @author [A. Rahardianto] [<veenone@gmail.com>]
+     * @since [v.6]
+     * @return int
+     */
+    public function checkoutTotal()
+    {
+        $checkedout = $this->users->count();
```
Comment (uberbrady) [inline app/Models/Consumable.php:356]: Doesn't this ignore the `$total` or `$totalnum` or whatever it is we're calling it?

File: app/Models/Consumable.php [hunk 10]
```diff
@@ -285,13 +286,87 @@ public function getEula()
...
+
+        
+    /**
+     * Gets the latest checkout note
+     *
+     * @author [A. Rahardianto] [<veenone@gmail.com>]
+     * @since [v.6]
+     * @return string
+     */
+    public function checkoutNote()
```
Comment (uberbrady) [inline app/Models/Consumable.php]: Similar to above, a better method name here would be very helpful - `lastCheckoutNote()` or something like that?

File: app/Notifications/CheckoutConsumableNotification.php [hunk 11]
```diff
@@ -123,6 +123,8 @@ public function toMail()
                 'eula'          => $eula,
                 'req_accept'    => $req_accept,
                 'accept_url'    => $accept_url,
+                'totalnum'      => $this->item->lastConsumed(),
+                'checkoutnote'  => $this->item->checkoutNote()
```
Comment (uberbrady) [inline app/Notifications/CheckoutConsumableNotification.php]: Ah! Now I see why you're doing the `lastConsumed()` and `checkoutNote()` methods above. Unfortunately, this could be subject to a race condition, in that if you did two checkouts of consumables-with-quantities at the exact same time, the wrong quantity and note might get picked up.

A better approach might be to instead fire the CheckoutConsumableNotification against the `users_consumables` record instead of specifying an Admin, and a Consumable, and a recipient. Or to at least include the quantity and checkoutnote in the constructor?

File: app/Services/PredefinedKitCheckoutService.php [hunk 12]
```diff
@@ -172,8 +172,10 @@ function () use ($user, $admin, $checkout_at, $expected_checkin, $errors, $asset
                         'consumable_id' => $consumable->id,
                         'user_id' => $admin->id,
                         'assigned_to' => $user->id,
+                        'totalnum' => $consumable->pivot->$totalnum,
```
Comment (uberbrady) [inline app/Services/PredefinedKitCheckoutService.php]: Are these `$` supposed to be here?

File: database/migrations/2013_11_25_013244_create_licenses_table.php [hunk 13]
```diff
@@ -0,0 +1,43 @@
+<?php
```
Comment (uberbrady) [inline database/migrations/2013_11_25_013244_create_licenses_table.php:1]: This file doesn't look like it belongs here.

File: public/web.config [hunk 14]
```diff
@@ -23,5 +23,12 @@
         <remove fileExtension=".woff2" />
         <mimeMap fileExtension=".woff2" mimeType="application/x-font-woff2" />
     </staticContent>
+        <directoryBrowse enabled="true" />
```
Comment (uberbrady) [inline public/web.config:26]: I don't think we want this part of the change actually

File: resources/lang/en-GB/admin/consumables/general.php [hunk 15]
```diff
@@ -8,4 +8,5 @@
     'remaining' 			            => 'Remaining',
     'total' 			                => 'Total',
     'update'                            => 'Update Consumable',
+    'note'                              => 'Remarks',
```
Comment (uberbrady) [inline resources/lang/en-GB/admin/consumables/general.php:11]: So these additions to the language files need to happen in `en`, so they then get added to CrowdIn, and then you can add the new translations to CrowdIn. See the docs here: https://snipe-it.readme.io/docs/translations

File: resources/lang/en/admin/consumables/message.php [hunk 16]
```diff
@@ -4,32 +4,40 @@
 
     'does_not_exist' => 'Consumable does not exist.',
 
-    'create' => [
+    'create' => array(
```
Comment (uberbrady) [inline resources/lang/en/admin/consumables/message.php]: If there's any way we can not include these unrelated changes, I would definitely appreciate it - makes it much easier to review.

File: resources/views/consumables/checkout.blade.php [hunk 17]
```diff
@@ -21,7 +21,7 @@
         @if ($consumable->id)
           <div class="box-header with-border">
             <div class="box-heading">
-              <h2 class="box-title">{{ $consumable->name }} </h2>
+              <h2 class="box-title">{{ $consumable->name }}   ({{ $consumable->numRemaining()  }} out of {{ $consumable->qty}}  {{ trans('admin/consumables/general.remaining') }}) </h2>
```
Comment (uberbrady) [inline resources/views/consumables/checkout.blade.php:24]: Unfortunately you'll have to translate `out of` as well. You might be able to cheat by using a symbol such as `/` ?

e.g. instead of "3 out of 5" something like "3 / 5 remaining"? Not sure that will look right though...

File: resources/views/consumables/checkout.blade.php [hunk 18]
```diff
@@ -67,16 +81,16 @@
               </div>
             @endif
           <!-- Note -->
-          <div class="form-group {{ $errors->has('note') ? 'error' : '' }}">
```
Comment (uberbrady) [inline resources/views/consumables/checkout.blade.php:70]: So we *do* have a `notes` field in consumables (it's not actually `note`, but that's not your bug!) We still probably want to display that note for the consumable - imagine it's for toner and the `notes` field might say "This toner is for Canon 12345 and 234567 laser-printers, and the 5xg-LX super-duper laser-printer" - that still needs to show up.

So I don't think we can just 'steal' the field for note here.

File: resources/views/consumables/checkout.blade.php [hunk 19]
```diff
@@ -67,16 +81,16 @@
...
-              <textarea class="col-md-6 form-control" id="note" name="note">{{ old('note', $consumable->note) }}</textarea>
-              {!! $errors->first('note', '<span class="alert-msg" aria-hidden="true"><i class="fas fa-times" aria-hidden="true"></i> :message</span>') !!}
+              <textarea class="col-md-6 form-control" id="checkoutnote" name="checkoutnote"></textarea>	      
+              {!! $errors->first('note', '<span class="alert-msg" aria-hidden="true"><i class="fa fa-times" aria-hidden="true"></i> :message</span>') !!}
             </div>
           </div>
         </div> <!-- .box-body -->
         <div class="box-footer">
-          <a class="btn btn-link" href="{{ URL::previous() }}">{{ trans('button.cancel') }}</a>
+          <a class="btn btn-link" href="{{ route('consumables.index') }}">{{ trans('button.cancel') }}</a>
```
Comment (uberbrady) [inline resources/views/consumables/checkout.blade.php]: This is _probably_ fine? But If we don't need to change this I'd rather not, there's enough other stuff already going on...

File: resources/views/consumables/view.blade.php [hunk 20]
```diff
@@ -22,7 +22,7 @@
       @if ($consumable->id)
       <div class="box-header with-border">
         <div class="box-heading">
-          <h2 class="box-title"> {{ $consumable->name }}</h2>
+          <h2 class="box-title"> {{ $consumable->name }}   ({{ $consumable->numRemaining()  }} out of {{ $consumable->qty}}  {{ trans('admin/consumables/general.remaining') }})</h2>
```
Comment (uberbrady) [inline resources/views/consumables/view.blade.php]: Another spot where we're saying "out of"

File: resources/views/partials/bootstrap-table.blade.php [hunk 21]
```diff
@@ -645,6 +645,11 @@ function cleanFloat(number) {
         return parseFloat(decimalfixed);
     }
 
+    
+    function totalFormatter(data) {
+        return 'Total : ' + data.length;
```
Comment (uberbrady) [inline resources/views/partials/bootstrap-table.blade.php]: This probably needs internationalization too. You can break into Blade syntax here if you need.

File: resources/views/users/view.blade.php [hunk 22]
```diff
@@ -705,6 +705,7 @@ class="table table-striped snipe-table table-hover"
         </div><!-- /accessories-tab -->
 
         <div class="tab-pane" id="consumables">
+          <h2>Consumable Checkout History</h2>
```
Comment (uberbrady) [inline resources/views/users/view.blade.php]: Probably needs internationalization too.

File: resources/views/users/view.blade.php [hunk 23]
```diff
@@ -726,19 +727,21 @@ class="table table-striped snipe-table table-hover"
...
+                  <th class="col-ld-8">{{ trans('general.notes') }}</th>
                 </tr>
               </thead>
               <tbody>
                 @foreach ($user->consumables as $consumable)
                 <tr>
                   <td>{!! $consumable->present()->nameUrl() !!}</td>
                   <td>
-                    {!! Helper::formatCurrencyOutput($consumable->purchase_cost) !!}
+                    {!! $consumable->pivot->totalnum !!}
```
Comment (uberbrady) [inline resources/views/users/view.blade.php]: This replaces the purchase cost with the pivot's totalnum, and that's probably not what we want to do.

File: resources/views/users/view.blade.php [hunk 24]
```diff
@@ -726,19 +727,21 @@ class="table table-striped snipe-table table-hover"
...
                   </td>
-                  <td>{{ $consumable->created_at }}</td>
+                  <td>{{ $consumable->pivot->created_at }}</td>
```
Comment (uberbrady) [inline resources/views/users/view.blade.php:781]: I think this is the right thing to do here actually - we're looking at a user's list of consumables, we don't care when the *consumable* was `created_at`, we care about when the actual consumable checkout occurred (which is, in fact, the Pivot's `created_at`). That seems pretty fair.

**Review by uberbrady — CHANGES_REQUESTED** [review]
Thanks for tackling this for us! If we can get a few fixes in I'm hoping that we can get this merged in. I think the general approach of having checkout notes and quantity in the pivot table definitely makes sense.

I think some careful renaming could be really useful here. We are calling a new value '$total` and `$totalnum` but it seems like that's just a checkout quantity. Could we call it something like `$checkout_qty` or something else like that instead? That'd make it a little clearer when we're reading this years and years and years in the future...

On some of the blades I'm seeing that you're re-purposing some of the visual space that we are already using; instead try and just be additive here (except for that one space where you pretty clearly found a bug; keep that fix in there :) )

As I mentioned in the review, you also included a lot of language files - which I definitely appreciate! But unfortunately, those changes will just get blown out when we next regenerate our data from CrowdIn, our internationalization platform. You'll just want to add strings for `en` and then once that gets imported into CrowdIn, we'd definitely love to have your translations for the other languages you know (which seems like a lot! Very cool :) )

I *do* love the "default 1" trick for the `totalnum` - that's a really clever way to get 'free' backwards-compatibility. Very smart.

If you can make the changes I'm requesting I'd love to get this merged. Thanks again for submitting it to us!

File: app/Events/CheckoutableCheckedOut.php (hunk 3 above)
Comment (veenone) [reply app/Events/CheckoutableCheckedOut.php:18]: ok I will review this

File: app/Models/Consumable.php (hunk 4 above)
Comment (veenone) [reply app/Models/Consumable.php]: ah.. typo.. I will update it

File: app/Models/Consumable.php (hunk 5 above)
Comment (veenone) [reply app/Models/Consumable.php]: to be updated

File: app/Models/Consumable.php (hunk 6 above)
Comment (veenone) [reply app/Models/Consumable.php]: noted

File: app/Models/Consumable.php (hunk 7 above)
Comment (veenone) [reply app/Models/Consumable.php:319]: this related to the usage checkoutConsumableNotification.php .  To be reviewed.

File: app/Models/Consumable.php (hunk 8 above)
Comment (veenone) [reply app/Models/Consumable.php]: Noted :)

File: app/Models/Consumable.php (hunk 9 above)
Comment (veenone) [reply app/Models/Consumable.php:356]: I will review back with the naming of these total related variables

File: app/Models/Consumable.php (hunk 10 above)
Comment (veenone) [reply app/Models/Consumable.php]: Noted

File: app/Notifications/CheckoutConsumableNotification.php (hunk 11 above)
Comment (veenone) [reply app/Notifications/CheckoutConsumableNotification.php]: I'm still wondering on the logic of parsing the information after checout to the email actually (pardon my limited knowledge on this topic 😅 ), is there any examples somewhere covering the same topic?

File: app/Services/PredefinedKitCheckoutService.php (hunk 12 above)
Comment (veenone) [reply app/Services/PredefinedKitCheckoutService.php]: to be fixed

File: database/migrations/2013_11_25_013244_create_licenses_table.php (hunk 13 above)
Comment (veenone) [reply database/migrations/2013_11_25_013244_create_licenses_table.php:1]: to be removed. I forgot to discard from my local repo

File: public/web.config (hunk 14 above)
Comment (veenone) [reply public/web.config:26]: maybe this change due to my local environment setup.. I will revert back to the initial condition

File: resources/lang/en-GB/admin/consumables/general.php (hunk 15 above)
Comment (veenone) [reply resources/lang/en-GB/admin/consumables/general.php:11]: noted. next time I will add only in en. do you think I should remove the one I made already in the other languages?

File: resources/lang/en/admin/consumables/message.php (hunk 16 above)
Comment (veenone) [reply resources/lang/en/admin/consumables/message.php]: noted, to be reverted

File: resources/views/consumables/checkout.blade.php (hunk 17 above)
Comment (veenone) [reply resources/views/consumables/checkout.blade.php:24]: noted , proper update to be considered

File: resources/views/consumables/checkout.blade.php (hunk 18 above)
Comment (veenone) [reply resources/views/consumables/checkout.blade.php:70]: got it, note to be displayed in conjunction the checkout note

File: resources/views/consumables/checkout.blade.php (hunk 19 above)
Comment (veenone) [reply resources/views/consumables/checkout.blade.php]: I updated this since I had a lot of checkouts and wanted to go back to the index by pressing back. but it keeps coming back to previous checkout processes that I did and it doesn't make any sense in my opinion.  maybe I'm missing the idea of the usage of the link, or I was abusing the link during my test.

To revert the implementation on this

File: resources/views/consumables/view.blade.php (hunk 20 above)
Comment (veenone) [reply resources/views/consumables/view.blade.php]: noted

File: resources/views/partials/bootstrap-table.blade.php (hunk 21 above)
Comment (veenone) [reply resources/views/partials/bootstrap-table.blade.php]: noted, to be fixed

File: resources/views/users/view.blade.php (hunk 22 above)
Comment (veenone) [reply resources/views/users/view.blade.php]: ah I missed this one.. will do

File: resources/views/users/view.blade.php (hunk 23 above)
Comment (veenone) [reply resources/views/users/view.blade.php]: after looking around other issues on the use case of this field, you're right.
I will update it by adding more column for it

File: resources/views/users/view.blade.php (hunk 24 above)
Comment (veenone) [reply resources/views/users/view.blade.php:781]: thanks, so this one can be resolved then?

File: resources/lang/en/admin/consumables/message.php (hunk 16 above)
Comment (veenone) [reply resources/lang/en/admin/consumables/message.php]: hmm strange.. after checking, mine is different with the one in develop projects.. maybe due to my initial mistake when first forking the project without fast forward. Fixed now

File: resources/views/consumables/checkout.blade.php (hunk 18 above)
Comment (veenone) [reply resources/views/consumables/checkout.blade.php:70]: the view is updated as below.
Logically, consumable note is not updateable, so I put the textarea to disable editing
image

**Comment by veenone** [comment]
Files have been updated according to the review. there's still 1 comment not yet put in resolved since I think I have a little bit different understanding. but this is easily resolved actually

File: resources/views/consumables/checkout.blade.php (hunk 19 above)
Comment (veenone) [reply resources/views/consumables/checkout.blade.php]: link reverted back in the last commit

**Comment by veenone** [comment]
everything should be according to review feedback now

**Comment by veenone** [comment]
@uberbrady , can you review may latest changes?

**Comment by whalehub** [comment]
@uberbrady Any update on this?

**Comment by snipe** [comment]
We're still reviewing the UX for this and how it relates to recent changes. If it gets merged (or closed), you'll be notified because you're following this thread.

**Comment by snipe** [comment]
Blergh - any chance you can resolve the conflicts? (I promise I won't take as long to review this time.)

**Comment by veenone** [comment]
Just saw your comment, sorry I also missed this action actually.
It's done now

**Comment by veenone** [comment]
I have resolved new conflicts 3 times so far. any chances you will review soon @snipe ?

**Comment by snipe** [comment]
This looks like it needs rebasing - I don't think you actually changed 36 files?

**Comment by mikeroq** [comment]
Shouldn't the field names on the pivot table be more like `total_num` and `checkout_note`? Or maybe named differently like `qty` and `note`?

At least if there is multiple words it should be in `snake_case` format. 

I'm not 100% what code syle this project follows but that's "standard" Laravel naming convention.

**Comment by snipe** [comment]
We typically use `first_last` if it's two words

**Comment by mikeroq** [comment]
The note upon checkout is no longer stored in the action log, thus removing it from the activity report. Seems like that should still be logged?

File: database/migrations/2013_11_25_013244_create_licenses_table.php (hunk 13 above)
Comment (mikeroq) [reply database/migrations/2013_11_25_013244_create_licenses_table.php:1]: Seems to still be here. I had to remove one of these to get migrations to run due to duplicate class names.

**Comment by veenone** [comment]
> This looks like it needs rebasing - I don't think you actually changed 36 files?

I just checked that I made changes on these files, mostly on some resource files for some languages (pardon my ignorance as this was my first PR and didn't know that I only need to update the english one :( )

additionally, I made some adjustments on the email template content as well to include the total number in the notification email

**Comment by veenone** [comment]
on the snake case format for some variables, should I make changes on them?  @mikeroq, can you point out on which part of pivot for me to update? and also for the note for action log one, can you point this one out as well ? I will include them

**Comment by mikeroq** [comment]
I'm just an observer ;) interested in your feature.

I just try to always name columns with multiple words like `assigned_to` instead of `assignedto`. Just follows best practices and makes for better readability. Same with variables/methods just depends on the code style. `$assignedTo or $assigned_to` vs `$assignedto`

**Comment by veenone** [comment]
ok, I have made the changes 😃 
- Enable checkout note on the activity log
image
- Renaming improper variable with snake_case as suggested
  - totalnum --> checkout_qty
  - checkoutnote --> checkout_note

**Comment by mikeroq** [comment]
Ok but now the note gets stored in two places, action log, and the pivot table. So I guess the real question is should that be duplicated or just stored in the action log.

Also not sure what the purpose of the checkout notes box is since it's serving the same purpose as the existing notes box (which is the checkout note anyways). Also did you check that components checkout controller?

**Comment by veenone** [comment]
Actually why I duplicated the notes box is to differentiate between the notes; which is in my understanding to contain the description / global information of the consumable, and then checkout_notes, which contains comment on each time admin checks out the consumable.

I'm checking the duplication now, but is there any concerns having the same information both in pivot & action log? and how do you quick check this pivot table?

**Comment by mikeroq** [comment]
This is the current checkout page ( I'm on mobile)
image
image

**Comment by veenone** [comment]
for mobile view, it uses different blade file? there's no quantity in your screenshot.
I think I need to look for it first..

**Comment by snipe** [comment]
We do not have different blades for desktop v mobile

**Comment by mikeroq** [comment]
> for mobile view, it uses different blade file? there's no quantity in your screenshot. I think I need to look for it first..

Those are screens from how it works currently, not your branch. I was just showing what the existing notes box does.

**Comment by veenone** [comment]
Ah okay, because I checked on my mobile my branch has shown the same layout already. So here I still keep the original notes for read only purpose as this will be only updated using edit and checkout note will be the one that always updateable on every checkout. They are stored on different table


Here notes on the activity log is displaying the checkout notes now

**Comment by veenone** [comment]
Regarding rebasing, should I rebase this PR to the last commit of develop branch @snipe ?

**Comment by whalehub** [comment]
@snipe requesting review

**Comment by whalehub** [comment]
@uberbrady requesting review

**Comment by ghost** [comment]
@snipe @uberbrady @mikeroq Is there anything we can do to help land this feature in Snipe-IT?

**Comment by uberbrady** [comment]
@whalehub I'm re-reviewing the PR now. Regarding rebasing @veenone - yes, it would need to be rebased against the latest develop. I'm hoping to follow-up with a more thorough review shortly.

File: app/Events/CheckoutableCheckedOut.php [hunk 25]
```diff
@@ -14,17 +14,25 @@ class CheckoutableCheckedOut
...
+    public $checkout_note;
+    public $total;
 
     /**
      * Create a new event instance.
      *
      * @return void
      */
-    public function __construct($checkoutable, $checkedOutTo, User $checkedOutBy, $note)
+    public function __construct($checkoutable, $checkedOutTo, User $checkedOutBy, $note, $checkout_qty=null, $checkout_note=null)
```
Comment (uberbrady) [inline app/Events/CheckoutableCheckedOut.php]: I wonder if this `$checkout_qty` default should be `1`?

File: database/migrations/2013_11_25_013244_create_licenses_table.php (hunk 13 above)
Comment (uberbrady) [inline database/migrations/2013_11_25_013244_create_licenses_table.php:1]: This one doesn't seem quite this file ought to be here?

File: database/migrations/2022_03_18_104400_add_total_to_consumables_users.php [hunk 26]
```diff
@@ -0,0 +1,38 @@
+<?php
```
Comment (uberbrady) [inline database/migrations/2022_03_18_104400_add_total_to_consumables_users.php:1]: This one either

**Review by uberbrady — CHANGES_REQUESTED** [review]
It's still a bit hard to parse what here is changing and what isn't - if I could get a fresh rebase I can take another look. I still like the idea; and I think we should do it. I think the architectural choices all seem solid. Thank you!

**Comment by veenone** [comment]
I'm sorry I haven't been able to check the project these past weeks.
I will try to rebase it then response to the request :)

File: app/Events/CheckoutableCheckedOut.php (hunk 25 above)
Comment (veenone) [reply app/Events/CheckoutableCheckedOut.php]: fixed

File: database/migrations/2013_11_25_013244_create_licenses_table.php (hunk 13 above)
Comment (veenone) [reply database/migrations/2013_11_25_013244_create_licenses_table.php:1]: removed from my branch

File: database/migrations/2022_03_18_104400_add_total_to_consumables_users.php (hunk 26 above)
Comment (veenone) [reply database/migrations/2022_03_18_104400_add_total_to_consumables_users.php:1]: this one is the migration script for my feature, so I keep it here

**Comment by veenone** [comment]
hi @uberbrady , I have rebased my branch.
I hope this rebase is properly done... looking forward for your review

**Comment by veenone** [comment]
Happy new year @snipe @uberbrady!
I hope this year one of my PRs can be reviewed and integrated .. 🤞

**Comment by whalehub** [comment]
@snipe @uberbrady Any update on this?

**Comment by ghanendaghestani** [comment]
Any idea when this feature/fix will be available. It is impacting our origination as well. Thanks

**Comment by rbarreiros** [comment]
Any update on this ? This would be a massive improvement!

**Comment by jacojazz** [comment]
Can't wait for this to be added!

**Comment by jayavman** [comment]
@snipe any chance of this getting merged soon?

**Comment by snipe** [comment]
This was handled a while back in a slightly different way.
